samples/boot/oauth2-integration/authorizationserver-custom-consent-page/src/test/java/sample/OAuth2AuthorizationServerCustomConsentPageApplicationTests.java

```diff
@@ -0,0 +1,161 @@
+		OAuth2Authorization authorization = OAuth2Authorization
+				.withRegisteredClient(registeredClientRepository.findByClientId("messaging-client"))
+				.principalName("user1")
+				.authorizationGrantType(AuthorizationGrantType.CLIENT_CREDENTIALS)
+				.attribute(OAuth2AuthorizationRequest.class.getName(),
+						OAuth2AuthorizationRequest.authorizationCode()
+								.authorizationUri("https://issuer.example.com").clientId("messaging-client")
+								.redirectUri(testConsentResultEndpoint)
+								.scopes(Stream.of("openid", "message.read", "message.write").collect(Collectors.toSet()))
+								.state("state")
+								.build())
+				.attribute(OAuth2ParameterNames.STATE, "state").build();
+
+		authorizationService.save(authorization);
+		when(mockAuthorizationConsentService.findById(any(), any())).thenReturn(null);
+	}
+
+	@Test
+	public void onConsentPage_ShouldBeAbleToConsentToAllScopes() throws IOException {
+		final HtmlPage consentPage = withDefaultUserLoggedIn(customConsentPageUrl);
+		assertThat(consentPage.getTitleText()).isEqualTo("Consent required");
+
+		List<HtmlCheckBoxInput> scopes = consentPage.querySelectorAll("input[name='scope']").stream()
+				.map(scope -> (HtmlCheckBoxInput) scope).collect(Collectors.toList());
+		for (HtmlCheckBoxInput scope : scopes) {
+			scope.click();
+		}
+
+		assertThat(scopes.stream().map(DomElement::getId).collect(Collectors.toList()))
+				.containsExactlyInAnyOrder("message.read", "message.write");
+		assertThat(scopes.stream().allMatch(HtmlCheckBoxInput::isChecked)).isTrue();
+
+		DomElement submitConsentButton = consentPage.querySelector("button[id='submit-consent']");
+		this.webClient.getOptions().setRedirectEnabled(false);
+
+		WebResponse approveConsentResponse = submitConsentButton.click().getWebResponse();
+		assertThat(approveConsentResponse.getStatusCode()).isEqualTo(HttpStatus.MOVED_PERMANENTLY.value());
+		String location = approveConsentResponse.getResponseHeaderValue("location");
+		assertThat(location).startsWith(testConsentResultEndpoint);
+		assertThat(location).contains("code=");
+	}
+
+	@Test
+	public void onConsentPage_ShouldBeAbleToCancelConsentApproval() throws IOException {
+		final HtmlPage consentPage = withDefaultUserLoggedIn(customConsentPageUrl);
+		assertThat(consentPage.getTitleText()).isEqualTo("Consent required");
+
+		DomElement cancelConsentButton = consentPage.querySelector("button[id='cancel-consent']");
+		this.webClient.getOptions().setRedirectEnabled(false);
+
+		WebResponse cancelConsentResponse = cancelConsentButton.click().getWebResponse();
+		assertThat(cancelConsentResponse.getStatusCode()).isEqualTo(HttpStatus.MOVED_PERMANENTLY.value());
+		String location = cancelConsentResponse.getResponseHeaderValue("location");
+		assertThat(location).endsWith("?error=access_denied&error_description=OAuth%202.0%20Parameter:%20client_id&error_uri=https://datatracker.ietf.org/doc/html/rfc6749%23section-4.1.2.1&state=state");
```

Comment:
I am not sure we should test against the very specific error URL. More specifically, I'm not 100% sure the correct `parameter` here should be `client_id`, so it may change in the future.

Just test against `error=access_denied`.